mmgen/core/evaluation/evaluation.py

```diff
@@ -252,49 +253,54 @@ def single_gpu_online_evaluation(model, data_loader, metrics, logger,
-        metric.num_images for metric in vanilla_metrics)
-    for metric in vanilla_metrics:
+    # define mmcv progress bar
+    max_num_images = 0
+    for metric in vanilla_metrics + probabilistic_metrics:
         metric.prepare()
-        # avoid print empty pbar for metrics do not need reals
-        if metric.num_real_feeded == metric.num_images:
-            continue
-        mmcv.print_log(f'Feed reals to {metric.name} metric.', 'mmgen')
-        pbar = mmcv.ProgressBar(metric.num_real_need)
-        # feed in real images
-        for data in data_loader:
-            # key for unconditional GAN
-            if 'real_img' in data:
-                reals = data['real_img']
-            # key for conditional GAN
-            elif 'img' in data:
-                reals = data['img']
-            else:
-                raise KeyError('Cannot found key for images in data_dict. '
-                               'Only support `real_img` for unconditional '
-                               'datasets and `img` for conditional '
-                               'datasets.')
-
-            if reals.shape[1] not in [1, 3]:
-                raise RuntimeError('real images should have one or three '
-                                   'channels in the first, '
-                                   'not % d' % reals.shape[1])
-            if reals.shape[1] == 1:
-                reals = reals.repeat(1, 3, 1, 1)
-            num_feed = metric.feed(reals, 'reals')
-            if num_feed <= 0:
-                break
+        max_num_images = max(max_num_images,
+                             metric.num_real_need - metric.num_real_feeded)
+    mmcv.print_log(f'Sample {max_num_images} real images for evaluation',
+                   'mmgen')
+    pbar = mmcv.ProgressBar(max_num_images)
+    for data in data_loader:
+        if 'real_img' in data:
+            reals = data['real_img']
+        # key for conditional GAN
+        elif 'img' in data:
+            reals = data['img']
+        else:
+            raise KeyError('Cannot found key for images in data_dict. '
+                           'Only support `real_img` for unconditional '
+                           'datasets and `img` for conditional '
+                           'datasets.')
 
-            pbar.update(num_feed)
+        if reals.shape[1] not in [1, 3]:
+            raise RuntimeError('real images should have one or three '
+                               'channels in the first, '
+                               'not % d' % reals.shape[1])
+        if reals.shape[1] == 1:
+            reals = reals.repeat(1, 3, 1, 1)
 
-        # finish the pbar stdout
-        sys.stdout.write('\n')
+        num_feed = 0
+        for metric in vanilla_metrics:
+            num_feed_ = metric.feed(reals, 'reals')
+            num_feed = max(num_feed_, num_feed)
+        for metric in probabilistic_metrics:
+            kwargs_ = deepcopy(kwargs)
+            kwargs_['mode'] = 'reconstruction'
```

Comment:
Just a suggestion. It seems that the probabilistic metrics are only used in reconstruction mode. Thus, we may rename it to `recon_metrics`. Another reason for this modification is that FID and IS indeed are also metrics based on distribution distance, which can also be regarded as probabilistic metrics.

Reply:
This suggestion is quite reasonable.